test/studies/shootout/binary-trees/binarytrees-pool.chpl

```diff
@@ -8,44 +8,45 @@
 
 use Allocators;
 
-config const n = 10,         // the maximum tree depth
-             globalPoolSize = (2 ** 31)-1,
-             localPoolSize = globalPoolSize;
-
-var globalPool = new bumpPtrMemPool(globalPoolSize, alignment=0);
+config const n = 10;                         // the maximum tree depth
 
 proc main() {
-  const minDepth = 4,                      // the shallowest tree
-        maxDepth = max(minDepth + 2, n),   // the deepest normal tree
-        strDepth = maxDepth + 1,           // the depth of the "stretch" tree
-        depths = minDepth..maxDepth by 2;  // the range of depths to create
-  var stats: [depths] (int,int);           // stores statistics for the trees
+  const minDepth = 4,                        // the shallowest tree
+        maxDepth = max(minDepth + 2, n),     // the deepest normal tree
+        strDepth = maxDepth + 1,             // the depth of the "stretch" tree
+        depths = minDepth..maxDepth by 2,    // the range of depths to create
+        nodeSize = 24,                       // the size of a node
+        poolSize = 2**(maxDepth+2)*nodeSize; // how much each pool allocates
+  var stats: [depths] (int,int);             // stores statistics for the trees
 
   //
   // Create the short-lived "stretch" tree, checksum it, and print its stats.
   //
   {
-    const strTree = newWithAllocator(globalPool, unmanaged Tree, strDepth, globalPool);
+    const pool = new bumpPtrMemPool(poolSize, alignment=0),
+          strTree = newWithAllocator(pool, unmanaged Tree, strDepth, pool);
     writeln("stretch tree of depth ", strDepth, "\t check: ", strTree.sum());
   }
 
   //
   // Build the long-lived tree.
   //
-  const llTree = newWithAllocator(globalPool, unmanaged Tree, maxDepth, globalPool);
+  const pool = new bumpPtrMemPool(poolSize, alignment=0),
+        llTree = newWithAllocator(pool, unmanaged Tree, maxDepth, pool);
 
   //
   // Iterate over the depths. At each depth, create the required trees in
   // parallel, compute their sums, and free them.
   //
   for depth in depths {
-    const iterations = 2**(maxDepth - depth + minDepth);
+    const iterations = 2**(maxDepth - depth + minDepth),
+          poolSize = iterations/here.maxTaskPar*2**(depth+1)*nodeSize;
```

Comment:
You're free to merge this anytime you're sick of fine-tuning it with me.  :)

Now that we have one poolSize used by two pools and another poolSize used by the others, what do you think about moving the computation of the poolSize into a helper routine, like:

```chapel
proc poolSize(depth, num=1) do return num * 2**(depth+1)*nodeSize;
```

and then the poolSize arguments in the `new bumpPtrMemPool` calls could become:

```chapel
…new bumpPtrMemPool(poolSize(maxDepth+1), alignment=0)…
const poolSize = poolSize(depth, iterations/here.maxTaskPar);
```

(though maybe that second one is too long to inline and still should be named and captured

Also, belated thought: Do we risk any issues due to `iterations/here.maxTaskPar` being a 'floor' division rather than 'ceil'?

Reply:
It definitely makes sense as a helper routine, with the one addition that I think it should be `inline`

> Also, belated thought: Do we risk any issues due to iterations/here.maxTaskPar being a 'floor' division rather than 'ceil'?

Yes 😭. `CHPL_RT_NUM_THREADS_PER_LOCALE=3 ./binary-trees-pool --n=21` and `CHPL_RT_NUM_THREADS_PER_LOCALE=5 ./binarytrees-pool --n=21` both demonstrate that problem. https://github.com/chapel-lang/chapel/pull/26009/commits/0135c15121899999646c47ad3cc6b93407b8bece uses `divCeilPos` to avoid that